include/openPMD/Series.hpp

```diff
@@ -282,5 +348,139 @@ class Series : public Attributable
+{
+    using iteration_index_t = IndexedIteration::index_t;
+
+    using maybe_series_t = auxiliary::Option< Series * >;
+
+    maybe_series_t m_series;
+    iteration_index_t m_currentIteration = 0;
+
+    // construct the end() iterator
+    SeriesIterator();
+
+public:
+    SeriesIterator( Series * );
+
+    SeriesIterator &
+    operator++();
+
+    IndexedIteration
+    operator*();
+
+    bool
+    operator==( SeriesIterator const & other ) const;
+
+    bool
+    operator!=( SeriesIterator const & other ) const;
+
+    static SeriesIterator
+    end();
+};
+
+/**
+ * @brief Reading side of the streaming API. Create instance via
+ *        Series::readIterations(). For use in a C++11-style foreach loop.
+ *        Designed to allow reading any kind of Series, streaming and non-
+ *        streaming alike. Calling Iteration::close() manually before opening
+ *        the next iteration is encouraged and will implicitly flush all
+ *        deferred IO actions. Otherwise, Iteration::close() will be implicitly
+ *        called upon SeriesIterator::operator++(), i.e. upon going to the next
+ *        iteration in the foreach loop.
+ *
+ *        Since this is designed for streaming mode, reopening an iteration is
+ *        not possible once it has been closed.
+ *
+ */
+class ReadIterations
+{
+    friend class Series;
+
+private:
+    using iterations_t = decltype( Series::iterations );
+    using iterator_t = SeriesIterator;
+
+    Series * m_series;
+
+    ReadIterations( Series * );
+
+public:
+    iterator_t
+    begin();
+
+    iterator_t
+    end();
+};
+
+/**
+ * @brief Writing side of the streaming API. Create instance via
+ *        Series::writeIterations().
+ *        For use via WriteIterations::operator[]().
+ *        Designed to allow reading any kind of Series, streaming and non-
+ *        streaming alike. Calling Iteration::close() manually before opening
+ *        the next iteration is encouraged and will implicitly flush all
+ *        deferred IO actions. Otherwise, Iteration::close() will be implicitly
+ *        called upon SeriesIterator::operator++(), i.e. upon going to the next
+ *        iteration in the foreach loop.
+ *
+ *        Since this is designed for streaming mode, reopening an iteration is
+ *        not possible once it has been closed.
+ *
+ */
+class WriteIterations : private Container< Iteration, uint64_t >
+{
+    friend class Series;
+
+private:
+    using iterations_t = Container< Iteration, uint64_t >;
+    struct SharedResources
+    {
+        iterations_t iterations;
+        auxiliary::Option< uint64_t > currentlyOpen;
+
+        SharedResources( iterations_t );
+        ~SharedResources();
+    };
+
+    using key_type = typename iterations_t::key_type;
+    using value_type = typename iterations_t::key_type;
+    WriteIterations( iterations_t );
+    explicit WriteIterations() = default;
+    // Index of the last opened iteration
+    std::shared_ptr< SharedResources > shared;
```

Comment:
Would `m_openIterationIndices` be a better name?

Reply:
Hm, my idea was that `WriteIterations` is a handle and that `SharedResources` is the internal class containing all members shared between instances. Which in this case, happen to be `iterations` and `currentlyOpen`, but that's coincidence ;)